crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -0,0 +1,100 @@
+use crate::semantic_index::ast_ids::HasScopedAstId;
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::expression::Expression;
+use crate::semantic_index::symbol::{ScopeId, ScopedSymbolId, SymbolTable};
+use crate::semantic_index::symbol_table;
+use crate::types::{infer_expression_types, IntersectionBuilder, Type, TypeInference};
+use crate::Db;
+use ruff_python_ast as ast;
+use rustc_hash::FxHashMap;
+use std::sync::Arc;
+
+/// Return type constraint, if any, on `definition` applied by `test`.
+pub(crate) fn narrowing_constraint<'db>(
+    db: &'db dyn Db,
+    test: Expression<'db>,
+    definition: Definition<'db>,
+) -> Option<Type<'db>> {
+    all_narrowing_constraints(db, test)
+        .get(&definition.symbol(db))
+        .copied()
+}
+
+#[salsa::tracked]
```

Comment:
You want `return_ref` here or salsa clones the value everytime the query is called
```suggestion
#[salsa::tracked(return_ref)]
```

Reply:
Surprisingly (to me) this didn't make a detectable difference in the benchmarks.

Reply:
I think perhaps that's just because in practice `NarrowingConstraints` will usually be quite small: either an empty `FxHashMap` or an `FxHashMap` with a single entry in it.

Reply:
Interesting. This is not for this PR, but it might be worth considering returning a `Vec` instead of a `HashSet` (or implementing our own `TinySet` that is based on `SmallVec`). Vecs are cheaper to create and can also be faster to search when they are only very few items.